CDS/cdsresponder/cdsresponder.rb

```diff
@@ -170,88 +194,105 @@ def ThreadFunc
+          #loop while not waiting for timeout
+          if @should_finish
+            puts "got should_finish signal"
+            @isexecuting = false
+            break
+          end
           File.delete(triggerfile)
           File.delete(@routefile)
         end #end block to catch exceptions
       }
+      #if should_finish is set while messages are NOT on the queue, then the poll times out without executing the loop break above.
+      #so we must re-do the check here.
+      if @should_finish
+        puts "got should_finish signal"
+        @isexecuting = false
+        break
+      end
     end #while @isexecuting
-
+    puts "reached end of threadfunc"
   end #def threadfunc
 
-  def join
-    @threadref.join
-    File.delete(@routefile)
+  #wait for the listener to terminate
+  # +limit+:: number of seconds to wait
+  # returns the thread id or nill if it timed out
+  def join(limit)
+    result = @threadref.join(limit)
+    File.delete(@routefile) if(@routefile)
+    result
   end #def join
 
+  #forcibly terminate the listener
+  def kill
+    @threadref.kill
+  end
+
 end #class CDSResponder
 
+def clean_shutdown(responders,server,timeout)
+  logger = Logger.new(STDOUT)
+  responders.each { |queue,resp|
+    puts "Shutting down listener for queue #{queue}"
+    resp.should_finish=true
+  }
+  responders.each { |queue,resp|
+    puts "Waiting for termination of listener for queue #{queue}"
+    if resp.join(timeout)==nil
+      puts "Queue listener for #{queue} failed to terminate after #{timeout} seconds, forcibly terminating"
+      resp.kill
+    end
+  }
+  puts "All listeners shut down, terminating server"
+  server.shutdown
+end
+
 ### START MAIN
 
-Raven.capture do
-  begin
-    #Process any commandline options
-    $options={:configfile => '/etc/cdsresponder.conf', :region => 'UNKNOWN'}
-    OptionParser.new do |opts|
-      opts.banner="Usage: cdsresponder.rb [--config=/path/to/config.file] [--region=aws-region]"
-
-      opts.on("-c", "--config CONFIGFILE", "Path to the configuration file.  This should contain the following:",
-              "   configuration-table={dynamodb table to use for configuration}",
-              "\troutes-table={dynamodb table to use for the routes content}",
-              "\tregion={AWS region to use for SQS and DynamoDB",
-              "\taccess-key={AWS access key} [Optional; default behaviour is to attempt connection via AWS roles",
-              "\tsecret-key={AWS secret key} [Optional; as above",
-              "\tsentry-dsn={Sentry DSN} [optional]",
-              "\tlogging-host={database host}",
-              "\tlogging-db={database-name}",
-              "\tlogging-user={database-user}",
-              "\tlogging-passwd={database-password}") do |cfg|
-        $options.configfile=cfg
-      end
+#Process any commandline options
+$options = Trollop::options do
```

Comment:
this is called only from the cdsresponder.conf upstart script, usually.  The arguments are no different so callers don't need to get updated in order to work